### ISSUE TITLE
Manipulating PDB files with Sire's new API

### ISSUE THREAD

**Issue by jmichel80** [OpenBioSim/sire#75]
Steve in my group has come across this problem. He’s trying to put together an OpenMM tutorial showing how to import a solvated protein-ligand complex in OpenMM to run an MD simulation.
He started by getting inputs from here

```
# Clone the repository
git clone https://github.com/MCompChem/fep-benchmark.git
# Take the last ligand from the hif2a benchmark
tail -n90 fep-benchmark/hif2a/ligands.sdf > PT2385.sdf
# Take the prepared protein structure
cp fep-benchmark/hif2a/5tbm_prepared.pdb .
receptor_path = "5tbm_prepared.pdb"
ligand_path = "PT2385.sdf"
```
He then adapted this [openff notebook](https://github.com/openforcefield/openff-toolkit/blob/stable/examples/toolkit_showcase/toolkit_showcase.ipynb) to generate via interchange openmm inputs

I asked him if he could also use[ this BSS node ](https://github.com/OpenBioSim/biosimspace/blob/devel/nodes/playground/BSSPrepNode.ipynb)

This doesn’t work because the pdb file _5tbm_prepared.pdb_ comes from Maestro and it contains a few atom and residue names that differ from those used in leap templates. Thus the BSS line in the above script 
 
`prot_p = BSS.Parameters.parameterise(protein, node.getInput("proteinff")).getMolecule()`
 
generates in the backend a leap process that fails due to the following issues.
 
 
```
FATAL:  Atom .R<ACE 1>.A<1H 7> does not have a type.
FATAL:  Atom .R<ACE 1>.A<2H 8> does not have a type.
FATAL:  Atom .R<ACE 1>.A<3H 9> does not have a type.
FATAL:  Atom .R<LEU 2>.A<H1 20> does not have a type.
FATAL:  Atom .R<HIE 11>.A<HD1 18> does not have a type.
FATAL:  Atom .R<HIE 76>.A<HD1 18> does not have a type.
FATAL:  Atom .R<NMA 112>.A<N 1> does not have a type.
FATAL:  Atom .R<NMA 112>.A<CA 2> does not have a type.
FATAL:  Atom .R<NMA 112>.A<H 3> does not have a type.
FATAL:  Atom .R<NMA 112>.A<HA1 4> does not have a type.
FATAL:  Atom .R<NMA 112>.A<HA2 5> does not have a type.
FATAL:  Atom .R<NMA 112>.A<HA3 6> does not have a type.
FATAL:  Atom .R<HOH 113>.A<O 1> does not have a type.
FATAL:  Atom .R<HOH 113>.A<H1 2> does not have a type.
FATAL:  Atom .R<HOH 113>.A<H2 3> does not have a type
FATAL:  Atom .R<HOH 114>.A<O 1> does not have a type.
FATAL:  Atom .R<HOH 114>.A<H1 2> does not have a type.
FATAL:  Atom .R<HOH 114>.A<H2 3> does not have a type.
FATAL:  Atom .R<HOH 115>.A<O 1> does not have a type.
FATAL:  Atom .R<HOH 115>.A<H1 2> does not have a type.
FATAL:  Atom .R<HOH 115>.A<H2 3> does not have a type.
FATAL:  Atom .R<HOH 116>.A<O 1> does not have a type.
FATAL:  Atom .R<HOH 116>.A<H1 2> does not have a type.
FATAL:  Atom .R<HOH 116>.A<H2 3> does not have a type.
```
 
The OpenFF notebook is able to deal with these issues because it uses [PDBFixer](https://github.com/openmm/pdbfixer) to rename atoms and residues in ACE, LEU and NMA using Amber naming conventions. The OpenFF Topology object is also able load a pdb that contains multiple molecules and is presumably clever enough to work out which type of histine it is dealing with on the basis of the atom names found. 
 
These issues can be fixed by running PDBFixer on the downloaded pdb, and then manually renaming HIE 11 and HIE 76 to HID 11, HID 76, and deleting the water molecules to end up with a single protein molecule. 
 
I thought I could put together a simple sire script to automate the above steps. That is
1)      Rename atom 1H in residue ACE 1 to H1 
2)      Rename atom 2H in residue ACE 1 to H2 
3)      Rename atom 3H in residue ACE 1 to H3 
4)      Rename atom H1 in residue LEU 2 to H
5)      Rename all atoms in residue HIE 11 to HID 11 
6)      Rename all atoms in residue HIE 76 to HID 76
7)      Move HOH 114, 115, 116 to new molecules.
8)      Save the updated protein, and the water molecules in separate pdb files. 
 
This would allow me to then use the updated protein pdb with the BSS parameterisation node. I could also update the BSS parameterisation node to also parameterise the crystallographic waters and add them to the system. 

I thought I could use cursors to select the atoms  (https://sire.openbiosim.org/tutorial/part03/02_cursors.html) . While I can select them easily it’s not clear if there is with the new API a simple way to rename atom names, rename residue names, delete atoms etc… ?  

Basically a few pointers on how this could be done would be great.

**Comment by lohedges** [comment]
Have you tried reading the PDB with [readPDB](https://biosimspace.openbiosim.org/api/generated/BioSimSpace.IO.readPDB.html#BioSimSpace.IO.readPDB), which can pre-process it with `pdb4amber` when `pdb4amber=True `. This doesn't work that well in my experience, so perhaps wrapping `pdbfixer` is a better solution.

**Comment by lohedges** [comment]
As it's a standalone Python tool from `conda-forge` there's no reason why we can't support it.

**Comment by jmichel80** [comment]
That would still leave a number of leap issues such as the number of molecules being greater than 1, and the HIS residue names. 

pdb4amber could work if the hydrogen atoms are deleted and added by this tool. I will check but it would be better to keep the original atom coordinates.

a weakness of BSS.parameterise is that it only works for leap ready PDB inputs (split in a single molecule). This means that users have to be very familiar with Amber tools (especially as we don’t currently return by default detailed error reports when leap has failed).

**Comment by lohedges** [comment]
Yes, it's a bit tricky since we've always wanted a user to prepare outside of BioSimSpace so that they specify the input exactly and we don't make any ambiguous choices. From the command-line, `pdb4amber` can do a lot more than we support via `readPDB`. I chose not to expose most options since some don't work reliably (the user would need to check that it worked) and some are experimental.

I'm happy to switch to `pdbfixer` if it's more flexible. I like that it's pure python and completely standalone. I'd rather we use existing tools rather than reinvent things ourselves, although it's probably easy to do in this case.

**Comment by lohedges** [comment]
As you say, regardless of `pdbfixer`, some manual renaming is necessary, so it would be good to be able to do this, or warn the user of the issue.

**Comment by chryswoods** [comment]
I'll show you how to do it soon ;-)

**Comment by lohedges** [comment]
Just to say that you can almost do it using `pdb4amber` as follows:

```sh
# First pass to remove water and non-standard residues.
pdb4amber --in 5tbm_prepared.pdb --dry --amber-compatible-residues --out tmp.pdb

# Second pass to add missing atoms.
pdb4amber --in tmp.pdb --add-missing-atoms --out 5tbm_pdb4amber.pdb
```

This results in the following error:

```
Created a new atom named: 1H within residue: .R<ACE 1>
Created a new atom named: 2H within residue: .R<ACE 1>
Created a new atom named: 3H within residue: .R<ACE 1>
Created a new atom named: H1 within residue: .R<LEU 2>
  total atoms in file: 1757
  The file contained 4 atoms not in residue templates
Checking Unit.

/home/lester/.conda/envs/openbiosim/bin/teLeap: Warning!
The unperturbed charge of the unit (-6.000000) is not zero.
FATAL:  Atom .R<ACE 1>.A<1H 7> does not have a type.
FATAL:  Atom .R<ACE 1>.A<2H 8> does not have a type.
FATAL:  Atom .R<ACE 1>.A<3H 9> does not have a type.
FATAL:  Atom .R<LEU 2>.A<H1 20> does not have a type.
```

It's annoying that _both_ `pd4amber` and `pdbfixer` can't generate suitable names for the hydrogens in ACE.

If you simply remove the hydrogens in the first step, then add missing atoms, then parameterisation works. However, the ACE residue doesn't include the atoms that were removed.

```sh
# First pass to remove water, hydrogens, and non-standard residues.
pdb4amber --in 5tbm_prepared.pdb --dry --noh --amber-compatible-residues --out tmp.pdb

# Second pass to add missing atoms.
pdb4amber --in tmp.pdb --add-missing-atoms --out 5tbm_pdb4amber.pdb
```

```python
In [1]: import BioSimSpace as BSS

INFO:numexpr.utils:Note: NumExpr detected 20 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
INFO:numexpr.utils:NumExpr defaulting to 8 threads.
WARNING:root:Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.

In [2]: m = BSS.IO.readMolecules("5tbm_pdb4amber.pdb")[0]

In [3]: mm = BSS.Parameters.ff14SB(m).getMolecule()

In [4]: for atom in mm.getResidues()[0]:
   ...:     print(atom)
   ...:
<BioSimSpace.Atom: name='H1', molecule=3 index=0>
<BioSimSpace.Atom: name='CH3', molecule=3 index=1>
<BioSimSpace.Atom: name='H2', molecule=3 index=2>
<BioSimSpace.Atom: name='H3', molecule=3 index=3>
<BioSimSpace.Atom: name='C', molecule=3 index=4>
<BioSimSpace.Atom: name='O', molecule=3 index=5>

In [5]: mm
Out[5]: <BioSimSpace.Molecule: number=3, nAtoms=1753, nResidues=111>

In [6]: original_m = BSS.IO.readMolecules("5tbm_prepared.pdb")[0]

In [7]: original_m
Out[7]: <BioSimSpace.Molecule: number=5, nAtoms=1770, nResidues=116>
```

Will have a play with `pdbfixer` too.

**Comment by chryswoods** [comment]
Here's a script that does all that you want.

The key parts are:

1. Deleting atoms is the same as searching for "not atoms" and then extracting the result. The `.extract()` function returns a molecule that contains only the atoms in a view. So `mol["not resname HOH"].extract()` will return a molecule that doesn't contain `HOH` residues.

2. The cursor has `.set_name()` and `.set_number()` functions that let you set the name and number of whatever is viewed. In the script below I show how to use these to change the names of all `HIS` residues to `HID` and how to rename the carbon atoms in the first residue so that they go from `C1` to `C3`

There is a little messiness in the script that I will clean up for `2023.4.0`. This messiness is that the extracted molecule has the same molecule number, so has to be renumbered if a molecule is split and recombined. Also, the `sr.save` function doesn't accept a python list of molecules yet, so we need to manually convert the list to a `SelectorMol`. I will fix this for `2023.3.0`. (there is also a small bug where renaming and renumbering a molecule with a cursor doesn't work, hence why I use the old API to do that for water - I will fix that for `2023.3.0`.

```
import sire as sr

# load a PDB file

mols = sr.load("3NSS")

# PDBs typically put everything into a single molecule
mol = mols[0]

# Water molecules are in residues called "HOH" - filter them out
print("Extracting the protein")
protein = mol["not resname HOH"].extract().molecule()

# We do want the water molecules though...
print("Extracting the water molecules")
water = mol["resname HOH"].extract().molecule()

print(f"Number of crystallographic waters is {len(water)}")

# These water molecules are just "O" atoms - break them into
# individual waters
waters = []

with sr.base.ProgressBar(len(water)) as bar:
    for atom in water.atoms():
        water = atom.extract().molecule()
        # we need to rename and renumber this molecule
        water = water.edit().rename("WAT").renumber().commit()
        waters.append(water)
        bar.tick()

print(f"Extracted {len(waters)} water molecules.") 

# Now we want to rename some of the atoms and residues in the protein.
# I will just do an example here of renaming all HIS residues to HID
cursor = protein.cursor()

for c in cursor["resname HIS"]:
    c.set_name("HID")

# Now I will rename the carbon atoms in the first residue
for i, c in enumerate(cursor["residx 0 and element C"]):
    c.set_name(f"C{i+1}")

protein = cursor.commit()

# save the protein to a PDB file
sr.save(protein, "cleaned_protein.pdb")

# save the waters to their own PDB file
sr.save(sr.mol.SelectorMol(waters), "cleaned_waters.pdb")
```

**Comment by chryswoods** [comment]
I am planning on making it easier to do molecular editing using the cursor API. Deleting atoms/residues etc will likely be by calling something like `cursor.remove()` or by calling `del cursor`.

I was delayed in this because I was trying to think of a nice way of letting people add atoms and residues. This is what pushed me to the openmm code, smiles / smarts etc. I now have the machinery to let people easily build and add fragments, perform energy minimisation etc. Also, with the smarts searching, I should be able to easily add templating support, letting users apply a set of residue templates to clean up molecules (e.g. standardise protein residue layouts / names, or even do residue mutations). These could auto-minimise using openmm.

**Comment by lohedges** [comment]
Actually, I was wrong, the ACE residue is correct. If you remove the waters using `pdb4amber` then use `pdbfixer` as follows, then it works, other than the `HIE` to `HID` issue.

```
pdbfixer 5tbm_prepared.pdb --replace-nonstandard --add-atoms=all --output 5tbm_pdbfixer.pdb
``` 

The files are the same other than the following:

```python
In [1]: import BioSimSpace as BSS

In [2]: m_pdb4amber = BSS.IO.readMolecules("5tbm_pdb4amber.pdb")[0]

In [3]: m_pdbfixer = BSS.IO.readMolecules("5tbm_pdbfixer.pdb")[0]

In [4]: for res0, res1 in zip(m_pdb4amber.getResidues(), m_pdbfixer.getResidues()):
   ...:     if res0.name() != res1.name():
   ...:         print(f"pdb4amber = {res0.name()}, pdbfixer = {res1.name()}")
   ...:
pdb4amber = HIE, pdbfixer = HIS
pdb4amber = HIE, pdbfixer = HIS
pdb4amber = HIE, pdbfixer = HIS
pdb4amber = HIE, pdbfixer = HIS
pdb4amber = HIE, pdbfixer = HIS
```

**Comment by chryswoods** [comment]
While waiting for the windows compile I cleaned up the niggles that made the above script a little messy. Here's a cleaned up script which will work with `2023.3.0` once it is released.

```
import sire as sr

# load a PDB file

print("Loading the protein from the PDB")
mols = sr.load("3NSS", silent=True)

# PDBs typically put everything into a single molecule
mol = mols[0]

# Water molecules are in residues called "HOH" - filter them out
print("\nExtracting the protein")
protein = mol["not resname HOH"].extract().molecule()

# We do want the water molecules though...
print("Extracting the water molecules")
water = mol["resname HOH"].extract().molecule()

print(f"Number of crystallographic waters is {len(water)}\n")

# These water molecules are just "O" atoms - break them into
# individual waters
waters = []

with sr.base.ProgressBar(total=len(water),
                         text="Dividing molecules:",
                         unit="molecules / second") as bar:
    for atom in water.atoms():
        waters.append(atom.extract().molecule())
        bar.tick()

    bar.success()

print(f"\nExtracted {len(waters)} water molecules.\n") 

# Now we want to rename some of the atoms and residues in the protein.
# I will just do an example here of renaming all HIS residues to HID
cursor = protein.cursor()

for c in cursor["resname HIS"]:
    c.set_name("HID")

# Now I will rename the carbon atoms in the first residue
for i, c in enumerate(cursor["residx 0 and element C"]):
    c.set_name(f"C{i+1}")

protein = cursor.commit()

# save the protein to a PDB file
print("Saving the protein...")
sr.save(protein, "cleaned_protein.pdb")

# save the waters to their own PDB file
print("Saving the waters...")
sr.save(waters, "cleaned_waters.pdb")

print("\nAll done")
```

**Comment by chryswoods** [comment]
I'll close this as the functionality is now in `2023.3.0` and the question is answered. Please feel free to reopen if needed.
